### PR TITLE
fix and improve empty template parameter pack handling

### PR DESCRIPTION
**Pull request by mitsuhiko**
This fixes a bug with various types of empty template parameter packs. This fixes compatibility with some weird cases that LLVM can deal with. This adds a few tests from llvm to the testsuite.

I also changed how the printing happens so that the commas are handled in an easier way that cover those new edge cases. Empty packs are retained in the AST.

### CODE REVIEW

**Comment by mitsuhiko** [comment]
cc @jauer for review.

File: src/lib.rs [hunk 1]
```diff
@@ -1182,6 +1182,7 @@ impl<'a> ParserState<'a> {
                 b'J' => Type::Int64(sc),
                 b'K' => Type::Uint64(sc),
                 b'W' => Type::Wchar(sc),
+                b'Q' => Type::Char(sc),
```
Comment (mitsuhiko) [inline src/lib.rs]: This is missing, I added it into the same PR and forgot about it. We can merge it together or I back this out.

**Comment by jrmuizel** [comment]
This looks reasonable to me.
